### PR TITLE
Feature

### PR DESCRIPTION
**Pull request by disc3110**
This project is a clubhouse where anyone can see the anonymous post but just singed in members can see who the author is. It's been written using Devise gem

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,24 +1,64 @@
-# README
+# Forms - Ruby on Rails
```
Comment (enaburekhan) [inline README.md]: Edit the title to reflect the current project. Which is `Members Only`.

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  end
+
+  # GET /posts/1 or /posts/1.json
+  def show
+    @comments = Comment.all
+  end
+
+  # GET /posts/new
+  def new
+    @post = current_user.posts.build
```
Comment (enaburekhan) [inline app/controllers/posts_controller.rb]: This should be: 
def new
  @post = Post.new
end

File: app/views/comments/_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,36 @@
+<%= form_with(model: @comment) do |form| %>
+    
+    <% if @comment.errors.any? %>
+    <div id="error_explanation">
+      <h2><%= pluralize(@comment.errors.count, "error") %> prohibited this post from being saved:</h2>
+
+      <ul>
+        <% @comment.errors.each do |error| %>
+          <li><%= error.full_message %></li>
+        <% end %>
```
Comment (enaburekhan) [inline app/views/comments/_form.html.erb]: Move the logic to a helper. There should be absolutely no logic  in view. Views are only use to render information that became available to them from it's corresponding controller and not for generating it.

File: app/views/layouts/_header.html.erb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+            <%= link_to 'Edit Profile', edit_user_registration_path, class:"nav-link" %>
+        </li>
+        <% else %>
+        <li class="nav-item">
+            <%= link_to 'Sign Up', new_user_registration_path, class:"nav-link" %>
+        </li>
+        <li class="nav-item">
+            <%= link_to 'Sign In', new_user_session_path, class:"nav-link" %>
+        </li>
+        <% end %>
```
Comment (enaburekhan) [inline app/views/layouts/_header.html.erb]: Also move this logic to a helper.

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -1,21 +1,32 @@
...
-    <%= yield %>
+  <body class="m-0">
 
-    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta3/dist/js/bootstrap.bundle.min.js" integrity="sha384-JEW9xMcG8R+pH31jmWH6WWP0WintQrMb4s7ZOdauHnUtxwoG2vI5DkLtS3qm9Ekf" crossorigin="anonymous"></script>
+    <%= render 'layouts/header' %>
+
+    <div class="container pt-1">
+    <% if notice %>
+    <%= render 'layouts/alerts' %>
+    <% end %>
```
Comment (enaburekhan) [inline app/views/layouts/application.html.erb]: Move this logic to a helper.

File: app/views/posts/_body.html.erb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+            <% end %>
+            <p class="card-text text-light"><%= post.body %></p>
+            <% if user_signed_in? && (current_user == post.user) %>
+            <%= link_to edit_post_path(post) do %>
+                <span class="icon"><i class="fas fa-edit"></i></span>
+            <% end %>
+            <%= link_to post, method: :delete, data: { confirm: 'Are you sure?' } do%>
+                <span class="icon"><i class="fas fa-trash-alt"></i></span>
+            <% end %>
+            <% end %>
```
Comment (enaburekhan) [inline app/views/posts/_body.html.erb]: Move this logic to a helper.

File: app/views/posts/_form.html.erb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    <div id="error_explanation">
+      <h2><%= pluralize(post.errors.count, "error") %> prohibited this post from being saved:</h2>
+
+      <ul>
+        <% post.errors.each do |error| %>
+          <li><%= error.full_message %></li>
+        <% end %>
+      </ul>
+    </div>
+  <% end %>
```
Comment (enaburekhan) [inline app/views/posts/_form.html.erb]: Move this logic to a helper.

File: app/views/posts/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+
+<div class="d-flex flex-column align-items-center">
+<% if user_signed_in? %>
+<%= render 'form', post: @post %>
+<% end %>
```
Comment (enaburekhan) [inline app/views/posts/index.html.erb]: Move this logic to a helper.

File: config/routes.rb [hunk 9]
```diff
@@ -1,3 +1,12 @@
 Rails.application.routes.draw do
+  resources :posts
+  resources :comments
+  devise_for :users, :controllers => { registrations: 'registrations' }
+
+  devise_scope :user do  
+    get '/users/sign_out' => 'devise/sessions#destroy'     
+ end
+
+  root 'posts#index'
```
Comment (enaburekhan) [inline config/routes.rb]: The application route should be created with only `resources :posts, only: [:new, :create, :index]`

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes required:hammer::wrench::full_moon:
Hi Diego!
Great job on this project:clap::clap:.

**Major highlights**

- The UI is awesome 
- A signed-in user can create new post and see post's author
- The application has Sign in/Sign out links present in any view of the app

However, we still need to make a few changes to get the project approved:muscle:.
**To improve**

- [ ] As highlighted in the PR below. Kindly edit the readme `title` to reflect the current project. Which is `Members Only. This will make it more descriptive.

- [ ] It creates the application routes using `resources :posts, only: [:new, :create, :index]` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_routes_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)). Kindly fix this.

- [ ] PostsController

- it uses `before_action` to restrict access to the `:new` and `:create` methods (uses ApplicationController helper method) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/before_action_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)). You can achieve this by creating a similar method like this below in the applicationController helper:

```ruby
def signed_in_only!
  # current_user in ApplicationController must be defined as helper method to make it work
  redirect_to  new_user_session_path unless current_user

  ## alternative solution:
  # if signed_in? is defined in ApplicationController (see stretch requirements)
  redirect_to new_user_session_path unless signed_in?
end
```
Then pass the below at the top of the Post controller:
`before_action :signed_in_only!, only: [:new, :create]`

- it initializes an empty post `@post = Post.new` in the `new` action. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_new_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)).

- [ ] As commented in the PR below. Kindly move all the highlighted logic in view to a helper. There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it. Check this [article](http://codefol.io/posts/Where-Do-I-Put-My-Code/) and this [article](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers) if you need help on how to use a helper in rails. see [screen-shot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from pathwright

**Optional**
- [ ] Kindly add author name to the readme.
- [ ] Please, cultivate the habit of committing messages often when building a project codebase. I expect to see more commit messages when fixing the changes.

Make the above-required changes and when you are ready, request another review.

Do have questions/comments? kindly drop them :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy hacking:blush:
Eric Enaburekhan.

**Review by smcommits — CHANGES_REQUESTED** [review]
# Status: Changes Required
## Hello, @disc3110! You have done a great job. Nice work implementing the changes that the previous TSE suggested. This project is almost complete. We just need address one criticl issue

## Required Changes

- [ ] **I cannot create a post. The application breaks. Please fix this**

I am assuming that the changes you made are causing this. Please find the issue and fix this. 
Application Breaks While Creating a Post
:------------------------------------------------------:
rails-new-createpost

# 
## Please reach out to me on Slack if you need assistance or I have overlooked something and made an unfair request.

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes required:wrench::hammer::full_moon:
**Hi Diego!
Great job so far:clap::clap:. Thank you for implementing almost all the required and optional changes:pray:. We are almost there. We just need to quickly fix one change to get the project approved:muscle:.**

**To improve**

- [ ] As I highlighted in my last review. Pass the `before_action :signed_in_only!, only: [:new, :create]` to the top of the post controller. To restrict access to the `new` and `create` methods. You have already defined the method in the application_helper in the controller. 

Make the above-required changes and when you are ready, request another review.

Do have questions/comments? kindly drop them :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy hacking:blush:
Eric Enaburekhan.

**Review by voscarmv — APPROVED** [review]
# Project status: Approved 💯 ✨ 😎 👍 

- Great work applying the last remaining details required by the previous code reviewer.

### You may merge this branch and move forward ⛙🛣️ 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻
# Happy Coding! 🧠 ⌨️ 🖱️ 🖥️

**Pull request merged**